src/chk/chk_internal.h

```diff
@@ -28,24 +28,137 @@
+#define X(a, b, c, d, e) a
 
 enum chk_rpc_opc {
-	CHK_PROTO_SRV_RPC_LIST
+	CHK_PROTO_SRV_RPC_LIST,
 	CHK_PROTO_SRV_RPC_COUNT,
 };
 
 #undef X
 
+/* check start in/out */
+#define DAOS_ISEQ_CHK_START							\
+	((uint64_t)		(csi_gen)		CRT_VAR)		\
+	((uint32_t)		(csi_flags)		CRT_VAR)		\
+	((int32_t)		(csi_phase)		CRT_VAR)		\
+	((d_rank_t)		(csi_leader_rank)	CRT_VAR)		\
+	((uint32_t)		(csi_padding)		CRT_VAR)		\
+	((d_rank_t)		(csi_ranks)		CRT_ARRAY)		\
+	((struct chk_policy)	(csi_policies)		CRT_ARRAY)		\
+	((uuid_t)		(csi_uuids)		CRT_ARRAY)
+
+#define DAOS_OSEQ_CHK_START							\
+	((int32_t)		(cso_status)		CRT_VAR)		\
+	((d_rank_t)		(cso_rank)		CRT_VAR)		\
+	((uint32_t)		(cso_phase)		CRT_VAR)		\
+	((uint32_t)		(cso_padding)		CRT_VAR)		\
+	((struct ds_pool_clue)	(cso_clues)		CRT_ARRAY)
+
+CRT_RPC_DECLARE(chk_start, DAOS_ISEQ_CHK_START, DAOS_OSEQ_CHK_START);
+
+/* check stop in/out */
+#define DAOS_ISEQ_CHK_STOP							\
+	((uint64_t)		(csi_gen)		CRT_VAR)		\
+	((uuid_t)		(csi_uuids)		CRT_ARRAY)
+
+#define DAOS_OSEQ_CHK_STOP							\
+	((int32_t)		(cso_status)		CRT_VAR)		\
+	((d_rank_t)		(cso_rank)		CRT_VAR)
+
+CRT_RPC_DECLARE(chk_stop, DAOS_ISEQ_CHK_STOP, DAOS_OSEQ_CHK_STOP);
+
+/* check query in/out */
+#define DAOS_ISEQ_CHK_QUERY							\
+	((uint64_t)		(cqi_gen)		CRT_VAR)		\
+	((uuid_t)		(cqi_uuids)		CRT_ARRAY)
+
+#define DAOS_OSEQ_CHK_QUERY							\
+	((int32_t)			(cqo_status)	CRT_VAR)		\
+	((int32_t)			(cqo_padding)	CRT_VAR)		\
+	((struct chk_query_pool_shard)	(cqo_shards)	CRT_ARRAY)
+
+CRT_RPC_DECLARE(chk_query, DAOS_ISEQ_CHK_QUERY, DAOS_OSEQ_CHK_QUERY);
+
+/* check mark in/out */
+#define DAOS_ISEQ_CHK_MARK							\
+	((uint64_t)		(cmi_gen)		CRT_VAR)		\
+	((d_rank_t)		(cmi_rank)		CRT_VAR)		\
+	((uint32_t)		(cmi_version)		CRT_VAR)
+
+#define DAOS_OSEQ_CHK_MARK							\
+	((int32_t)		(cmo_status)		CRT_VAR)
+
+CRT_RPC_DECLARE(chk_mark, DAOS_ISEQ_CHK_MARK, DAOS_OSEQ_CHK_MARK);
+
+/* check act in/out */
+#define DAOS_ISEQ_CHK_ACT							\
+	((uint64_t)		(cai_gen)		CRT_VAR)		\
+	((uint64_t)		(cai_seq)		CRT_VAR)		\
+	((uint32_t)		(cai_cla)		CRT_VAR)		\
+	((uint32_t)		(cai_act)		CRT_VAR)		\
+	((uint32_t)		(cai_flags)		CRT_VAR)
+
+#define DAOS_OSEQ_CHK_ACT							\
+	((int32_t)		(cao_status)		CRT_VAR)
+
+CRT_RPC_DECLARE(chk_act, DAOS_ISEQ_CHK_ACT, DAOS_OSEQ_CHK_ACT);
+
+/* check report in/out */
+#define DAOS_ISEQ_CHK_REPORT							\
+	((uint64_t)		(cri_gen)		CRT_VAR)		\
+	((uint32_t)		(cri_ics_class)		CRT_VAR)		\
+	((uint32_t)		(cri_ics_action)	CRT_VAR)		\
+	((int32_t)		(cri_ics_result)	CRT_VAR)		\
+	((d_rank_t)		(cri_rank)		CRT_VAR)		\
+	((uint32_t)		(cri_target)		CRT_VAR)		\
+	((uint32_t)		(cri_padding)		CRT_VAR)		\
+	((uuid_t)		(cri_pool)		CRT_VAR)		\
+	((uuid_t)		(cri_cont)		CRT_VAR)		\
+	((daos_unit_oid_t)	(cri_obj)		CRT_VAR)		\
+	((daos_key_t)		(cri_dkey)		CRT_VAR)		\
+	((daos_key_t)		(cri_akey)		CRT_VAR)		\
+	((d_string_t)		(cri_msg)		CRT_VAR)		\
+	((uint32_t)		(cri_options)		CRT_ARRAY)		\
+	((d_sg_list_t)		(cri_details)		CRT_ARRAY)
```

Comment:
an array of d_sg_list_t, all transfer by RPC inline? assume the size is not large? (or use bulk if needed).

Reply:
It is expected very small information for that.